Query vectorization.

gianm opened the pull request:
Implementation of the proposal in #7093.

drcrallen commented:
Very cool, I think both @xvrl and @nishantmonu51 may have taken a hack at this previously as well. There is even https://github.com/metamx/druid/tree/gpudruid if anyone wants to peek at super old code. Looking forward to see how this patch performs

gianm commented:
There's some benchmarks in the description (https://static.imply.io/gianm/vb.html) & I have also run some queries on a real cluster and seen similar speedups. Although notably, topN isn't vectorized in this patch, and contributions are welcome on that one 😄. I would expect speedup ratios there similar to groupBy, maybe a bit less since topN already has some optimizations targeted at reducing method call overhead (the monomorphic specialization system).

There's some TeamCity inspections going bump because of unused stuff. Most of it is there for a reason so I'll add some comments that hopefully make the inspections happy.

gianm commented:
Updated the top comment to be more proposal-y following discussion on the dev list.

b-slim commented:
> Updated the top comment to be more proposal-y following discussion on the dev list.

I do not think this is going to work, because the updates to the description is something that only authors are doing and there is no tracking for history.
Also take the example of Rejected approach, not sure whom had rejected that ? and whom thought does reflects.
In fact, i might not agree that Apache Arrow is not a good candidate, in fact am looking to see if we can use it to stream bulk data from Druid and expose it a standard interface to export data to another execution engine...

gianm commented:
> I do not think this is going to work, because the updates to the description is something that only authors are doing and there is no tracking for history.

There is history (click "edited" in the top bar of the comment). I don't think there's anything wrong with edits coming from the author of the proposal, so long as they reflect community discussion.

> Also take the example of Rejected approach, not sure whom had rejected that ? and whom thought does reflects.

I used the Kafka template from https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals. As I understand it, this section means "rejected by the author of the proposal", and is meant to help reflect the author's thinking in the course of developing the proposal. In the current version it reflects my thinking alone. I would update it to reflect community discussion.

> In fact, i might not agree that Apache Arrow is not a good candidate, in fact am looking to see if we can use it to stream bulk data from Druid and expose it a standard interface to export data to another execution engine...

That sounds like a good discussion point!! Would love to hear any more thoughts you have on this topic. As I see it- supporting Arrow as an export interface and using it internally are two different things and could happen independently of each other. There is also a potential case to be made around using Arrow for the later stages of query execution, after per-segment processing, even if it's not used for per-segment processing. Druid doesn't currently always use optimized data structures for this phase (there's a lot of in-heap Java objects involved, and some glaring inefficiencies like representing rows using LinkedHashMaps).

Fwiw, the KIP process suggests the following for discussion,

> Start a [DISCUSS] thread on the Apache mailing list. Please ensure that the subject of the thread is of the format [DISCUSS] KIP-{your KIP number} {your KIP heading} The discussion should happen on the mailing list not on the wiki since the wiki comment system doesn't work well for larger discussions. In the process of the discussion you may update the proposal. You should let people know the changes you are making.

If we want to try something similar -- we can do that in comments for this PR.

Do you think that will work?

gianm commented:
Some previous discussion on Arrow as an export format in particular: https://github.com/apache/incubator-druid/issues/3891

jihoonson commented:
> There is history (click "edited" in the top bar of the comment). I don't think there's anything wrong with edits coming from the author of the proposal, so long as they reflect community discussion.

This is true, but I would like to suggest to use a better tool for easy discussion. I think the proposal will be usually large, which means it's not easy to point some particular parts of the proposal. Probably we can use Github PR for the proposals as well, so that people can easily leave comments.

b-slim commented:
> If we want to try something similar -- we can do that in comments for this PR.

As i said i am not targeting per-say  this PR. Ideally would love to see a formal agreement about the process on the dev mailing list and then maybe start implementing it on this PR if possible.

> There is history (click "edited" in the top bar of the comment). I don't think there's anything wrong with edits coming from the author of the proposal, so long as they reflect community discussion.

Thanks i guess i learned new thing today.

FYI as a side note I honestly  prefer a separation between Code reviewing and Proposal reviewing discussion , for the simple reason that when the code review starts it will introduce a lot of noise about the implementation details and therefore all the discussion about design will be lost or not possible to sort out, In fact Github pages becomes unloadable once we get bunch of comments.

gianm commented:
Sounds good, let's take that part of the discussion to the mailing list.

gianm commented:
I added a "Future work" section that illustrates what would be left to be done before the feature is 'complete'. I don't necessarily think it makes sense to do any of it in the first PR, since the first PR would just be to lay a foundation that allows the other work to happen.

fjy commented:
👍 on this proposal - looks to be a win across the board

b-slim commented:
@gianm do you want to keep the discussion here or start an issue as design review?

b-slim commented:
Also i saw this `org.apache.druid.segment.incremental.IncrementalIndexStorageAdapter.IncrementalIndexCursor#advanceUninterruptibly` was removed can someone explain to me what is used for?

b-slim commented:
wondering why we need to have `org.apache.druid.segment.vector.VectorColumnSelectorFactory#getMaxVectorSize` isn't enough to use the cursor one ?

b-slim commented:
> Vectorizing other query engines: topN, scan, select, search.

Do we really need to support select? IMO we need to drop it as soon as we can it causes lot of OOM and i don't see why would use it when you can use scan.

gianm commented:
@b-slim,

> @gianm do you want to keep the discussion here or start an issue as design review?

I was waiting to see how the thread on the dev list goes and then do what we decide there.

> Noticed that advanceTo(offset) is dropped from the Cursor API, am wondering if you think it is not needed or because the current implem is a synthetic sugar over advance() ?

I wasn't sure how to translate that specific API to the vectorized impl (it doesn't necessarily make sense to seek to a _specific_ offset when you're talking about blocks of rows) & thought it made more sense for the query engines that need it to do it themselves. It's only used in one place by the select query.

> Do you think the VectorCursor Api for Realtime based segments is going to be the same? or you have something else in mind?

It should be the same, I think, just like the regular Cursor is the same. You would expect vectorization to not have as much of an impact on IncrementalIndex, because bulk column reads don't improve efficiency the same way they do on a QueryableIndex, but you might still see some benefit on other parts of the query pipeline.

> Also i saw this org.apache.druid.segment.Cursor#advanceUninterruptibly and org.apache.druid.segment.Cursor#isDoneOrInterrupted are removed can someone explain to me what is used for? and why we do not need it for Vector APIs

They're used for topN prototypes & were added in #3889. I'm not sure if we'll want vectorized analogs or not. I am thinking that the topN prototype system is probably not _as_ necessary for a potentially vectorized topN engine, since part of the point of monomorphic processing is to reduce virtual method call overhead, but vectorization does that too (in a different way).

> wondering why we need to have org.apache.druid.segment.vector.VectorColumnSelectorFactory#getMaxVectorSize isn't enough to use the cursor one ?

I think it ended up being used by AggregatorFactories, which get a column selector factory when they factorize an aggregator, but not a cursor. It ended up being useful to add max/current vector size getters to a variety of places for reasons like that.

> Do we really need to support select? IMO we need to drop it as soon as we can it causes lot of OOM and i don't see why would use it when you can use scan.

#6088 is the only real reason to use select. I think when we have a solution to that problem we should definitely deprecate & drop it. I could see the potential to never vectorize it, and favor improving scan -> dropping select instead. It'd just be important to vectorize all query engines that exist at some point in time, if we want to get rid of the non-vectorized code paths.

fjy commented:
👍 

I am submitting my vote after a cursory review, mainly because of the author. Disclaimer, we also do work together.

@gianm has architected and developed significant components to Druid including the indexing service, SQL support, groupBy v2, expressions, and scan query, all of which have been changes as large as this one and all of which have worked near flawlessly in production, all while unlocking future key future features that will take Druid to the next level. He is also the primary architect on other major features including Druid's security framework, revamped batch ingestion, improved histogram support, system tables, greatly improved docs, and much more. I can't name another active committer with anywhere near the same level of contribution and track record. That is enough validation for my vote.

gianm commented:
Ha, thanks @fjy for the vote of confidence. I need to copy and paste that into a LinkedIn reference.

I am still looking forward to a robust discussion. I updated the "Proposed change" section with some more high level details in an effort to make the proposed change more clear without a need to study the code.

gianm commented:
Some discussion of how filtered aggregators work in this proposed change has come up in https://github.com/apache/incubator-druid/issues/6889#issuecomment-457073251.

sascha-coenen commented:
Great PR!!

I do not want to go off topic, but with regard to: 
>> Benchmarks on the code in this PR indicate we could get a 1.3-3x speedup

..., I meant to mention that there is an Israeli startup which is specialising in FPGA accelerations. 
With AWS F1 instance types being a thing now, FPGA might finally be getting some more traction. This startup has customised Druid's Query Engine to delegate the aggregation to FPGA hardware and they claim a 35x speedup for mostly-in-memory use-cases.

They might be an interesting technology partner in validating the design of a vectorisation layer and providing an FPGA based extension for it. I'm not affiliated with them in any way, so this is not me advertising or pitching them. I just found it interesting what they are doing and was curious as to whether you guys have heard of them and whether this sort of stuff would be a fit for what you are trying to achieve with this PR.

egor-ryashin commented on processing/src/main/java/org/apache/druid/query/filter/vector/VectorValueMatcher.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * A vectorized value matcher.
+ *
+ * @see org.apache.druid.query.filter.ValueMatcher, the non-vectorized version
```
ValueMatcher doesn't have a description and ValueMatcher list of implementation classes is quite big to understand its true role clearly, I think there should be a reference to a thorough description with a showcase code explaining (in general) the supposed way of interaction of ValueMatcher, Filter, Cursor and possibly other entities. And I think all other interfaces in that chain should also contain that reference link as most of them also have no description while having a huge list of implementations.

gianm replied:
I added some more javadocs to ValueMatcher and VectorValueMatcher that should hopefully point readers in the right direction. Its main role is to power filtered cursors (the ones that don't use indexes, anyway) and filtered aggregators.

Fwiw, if you want to learn more about the Druid query path, the best way IMO is to try starting at QueryableIndexStorageAdapter's `makeCursors` method and spider out from there to check out what it does and what all of its callees do. That only covers the immutable-segment part of the query path (mutable realtime segments go through IncrementalIndexStorageAdapter), but the immutable-segment code is where more of the action happens anyway, so it's a better place to start. After that it is also useful to look at the callers of `makeCursors` (generally query engines) and see what they do. Usually the first thing they do is make selectors and aggregators, then start walking through and processing rows. Even if javadocs on all these interfaces were improved, this exercise would still be useful, since it'd better show how everything ties together.

gianm commented:
@sascha-coenen,

> I meant to mention that there is an Israeli startup which is specialising in FPGA accelerations. With AWS F1 instance types being a thing now, FPGA might finally be getting some more traction. This startup has customised Druid's Query Engine to delegate the aggregation to FPGA hardware and they claim a 35x speedup for mostly-in-memory use-cases.
>
> They might be an interesting technology partner in validating the design of a vectorisation layer and providing an FPGA based extension for it. I'm not affiliated with them in any way, so this is not me advertising or pitching them. I just found it interesting what they are doing and was curious as to whether you guys have heard of them and whether this sort of stuff would be a fit for what you are trying to achieve with this PR.

It sounds cool, I haven't heard of them. I searched around for a bit and didn't see anything do you have a link? My feeling is that it is useful to optimize performance on CPU (what this patch does) regardless of the existence of specialized hardware. Specialized hardware is still quite interesting though in its own right. And if optimizations geared towards improving performance on CPUs help on other types of hardware too, all the better.

gianm commented:
I have moved the proposal to #7093, used the new proposal template, and added a lot of info, including tables detailing changes to interfaces and calling out differences between standard and vectorized versions.

egor-ryashin commented on processing/src/main/java/org/apache/druid/query/timeseries/TimeseriesQueryEngine.java:
```diff
@@ -19,24 +19,58 @@
+  /**
+   * Constructor for tests. In production, the @Inject constructor is used instead.
+   */
+  public TimeseriesQueryEngine()
```
`@VisibleForTesting` ?

egor-ryashin replied:
@gianm I have received a review request, I suspect you missed this and the following comments.

gianm replied:
Oh, I did miss this first set of comments. I'll take another look.

gianm replied:
Did you have any further comments, btw, other than these?

egor-ryashin replied:
Nope, looks good overall.

gianm replied:
@egor-ryashin Thanks! I just pushed up a new commit addressing your comments.

egor-ryashin commented on processing/src/main/java/org/apache/druid/query/filter/vector/ReadableVectorMatch.java:
```diff
@@ -0,0 +1,66 @@
+   * Even though this array is technically mutable, it is very poor form to mutate it if you are not the owner of the
+   * VectorMatch object.
+   */
+  int[] getSelection();
```
I got it, with _getSelection_ we expecting to trigger **HotSpot vectorization** elsewhere, that way we could leave the method but I'd indicate the purpose in the method comment.

gianm replied:
Yep, that is it. I added a comment.

egor-ryashin replied:
Could you mention `HotSpot vectorization` term in the comment?

gianm replied:
I added this sentence:

> Potential optimizations could include making it easier for the JVM to use CPU-level vectorization, avoid method calls, etc.

I'm not 100% sure that the CPU level vectorization will _actually_ be triggered, but it could theoretically be, which is the point. Even if it isn't today, it might be in future JVMs. (And it might be today, I just haven't checked.)

richardstartin commented:
The batch iterators used here use a lot less memory per bitmap iteration as of RoaringBitmap 0.8.2 [(see here)](https://github.com/RoaringBitmap/RoaringBitmap/pull/329/).

himanshug commented:
/subscribe

gianm commented on processing/src/main/java/org/apache/druid/segment/data/CompressedVSizeColumnarMultiIntsSupplier.java:
```diff
@@ -38,17 +38,21 @@
  * offsets - {@link ColumnarInts} of length num of rows + 1 representing offsets of starting index of first element of
- *           each row in values index and last element equal to length of values column, the last element in the offsets
- *           represents the total length of values column.
+ * each row in values index and last element equal to length of values column, the last element in the offsets
```
It does sound like the same thing. I actually didn't mean to change anything here, it was just IDE autoformatting. I restored it.

gianm commented on processing/src/main/java/org/apache/druid/segment/data/BlockLayoutColumnarFloatsSupplier.java:
```diff
@@ -105,17 +106,59 @@ public float get(int index)
-            )
-        );
+      // division + remainder is optimized by the compiler so keep those together
+      int bufferNum = start / sizePer;
```
I just added a comment to the declaration. I'm trying to keep the size of the diff as small as possible, it's already quite long even without renames like this.

gianm commented on processing/src/main/java/org/apache/druid/query/timeseries/TimeseriesQueryEngine.java:
```diff
@@ -45,65 +79,210 @@
+                      return null;
+                    }
+
+                    final TimeseriesResultBuilder bob = new TimeseriesResultBuilder(
```
Bob the builder :)

![image](https://user-images.githubusercontent.com/1214075/60709502-6f9fcb80-9ec5-11e9-8e79-07e9ce7fbe89.png)

gianm commented on processing/src/main/java/org/apache/druid/query/vector/VectorCursorGranularizer.java:
```diff
@@ -0,0 +1,171 @@
+      final StorageAdapter storageAdapter,
+      final VectorCursor cursor,
+      final Granularity granularity,
+      final Interval cursorInterval
```
I renamed it to `queryInterval`.

gianm commented on processing/src/main/java/org/apache/druid/query/vector/VectorCursorGranularizer.java:
```diff
@@ -0,0 +1,171 @@
+  {
+    final DateTime minTime = storageAdapter.getMinTime();
+    final DateTime maxTime = storageAdapter.getMaxTime();
+    final Interval actualInterval = cursorInterval.overlap(new Interval(minTime, granularity.bucketEnd(maxTime)));
```
I thought about it a bit and I think `clippedQueryInterval` is the clearest name. I used that one.

gianm commented on processing/src/main/java/org/apache/druid/query/timeseries/TimeseriesQueryEngine.java:
```diff
@@ -45,65 +79,210 @@
+      final TimeseriesQuery query,
+      final StorageAdapter adapter,
+      @Nullable final Filter filter,
+      final Interval interval,
```
Sure.

gianm commented on processing/src/main/java/org/apache/druid/query/timeseries/TimeseriesQueryEngine.java:
```diff
@@ -45,65 +79,210 @@
 
-    final Filter filter = Filters.convertToCNFFromQueryContext(query, Filters.toFilter(query.getDimensionsFilter()));
+    final Filter filter = Filters.convertToCNFFromQueryContext(query, Filters.toFilter(query.getFilter()));
+    final Interval interval = Iterables.getOnlyElement(query.getIntervals());
```
I added a javadoc noting this to the `process` method.

gianm commented:
The Travis errors are interesting; the tests run OOM. I got a heap dump and saw that ~78% of the heap is being used by org.apache.maven.plugin.surefire.report.WrappedReportEntry objects. It sounds like https://issues.apache.org/jira/browse/SUREFIRE-1147. It seems like there are too many tests now for surefire to handle!

I'll try updating to the latest 2.x surefire (2.22.2). If that doesn't change anything I'll try bumping up the Xmx in Travis a bit. Right now it's set to 512m. If _that_ doesn't work then I guess we either need to patch surefire, or refactor the tests to not have so many test cases.

gianm commented:
Updating surefire didn't help, but I kept the new version anyway. I'll try bumping Xmx to 800m.

gianm commented:
It looks like that bump worked… for now :)

gianm commented on processing/src/main/java/org/apache/druid/segment/QueryableIndexCursorSequenceBuilder.java:
```diff
@@ -0,0 +1,618 @@
+   *
+   * @return index of timestamp, or negative number equal to (-(insertion point) - 1).
+   */
+  private static int timeSearch(
```
I added tests to QueryableIndexCursorSequenceBuilderTest.

gianm commented on processing/src/main/java/org/apache/druid/segment/QueryableIndexCursorSequenceBuilder.java:
```diff
@@ -0,0 +1,618 @@
+
+public class QueryableIndexCursorSequenceBuilder
+{
+  // At this threshold, timestamp searches switch from binary to linear. The idea is to avoid too much decompression
```
Sure, I moved it to the method-level javadoc.

gianm commented on processing/src/main/java/org/apache/druid/segment/QueryableIndexCursorSequenceBuilder.java:
```diff
@@ -0,0 +1,618 @@
+      }
+
+      final int result = timeSearch(timestamps, interval.getEndMillis(), startOffset, index.getNumRows());
+      if (result >= 0) {
```
I just chose it so `timeSearch` would work similarly to `Arrays.binarySearch`. However, I guess there's no point in doing this, so I moved this block inside `timeSearch` and changed its contract appropriately.

egor-ryashin replied:
Hmm, looks like this block is still there:
![image](https://user-images.githubusercontent.com/14215045/60973727-7c8f4580-a331-11e9-96d5-abf521511580.png)

gianm replied:
Ah, you're right. It is dead code now, since the new `timeSearch` method always returns something >= 0. I removed it.

gianm commented on benchmarks/src/main/java/org/apache/druid/benchmark/query/SqlBenchmark.java:
```diff
@@ -64,39 +60,112 @@
-  private GroupByQuery groupByQuery;
-  private String sqlQuery;
-  private Closer resourceCloser;
+  private static final List<String> QUERIES = ImmutableList.of(
```
That all sounds good. More realistic benchmark datasets would be great. Maybe we could even use a realistic data generator like https://github.com/joke2k/faker (if there is something like this available in Java).

gianm commented on docs/content/querying/query-context.md:
```diff
@@ -60,3 +60,31 @@ In addition, some query types offer context parameters specific to that query ty
+
+|property|default| description|
+|--------|-------|------------|
+|vectorize|`false`|Enables or disables vectorized query execution. Possible values are `false` (disabled), `true` (enabled if possible, disabled otherwise, on a per-segment basis), and `force` (enabled, and groupBy or timeseries queries that cannot be vectorized will fail). The `"force"` setting is meant to aid in testing, and is not generally useful in production (since real-time segments can ne ...
```
The idea is that if you set `force` then you know your query will be 100% vectorized. It's meant to aid in testing and development -- if you want to be totally sure you're testing vectorized code paths. `true` is the setting that is meant to be useful in production scenarios.

By the way, I don't see a reason to _not_ eventually vectorize processing of the IncrementalIndex. I'm not sure if it'll give a speed boost, but it might allow us to eventually delete the nonvectorized code paths. Maintaining both forever would be a pain.

gianm commented on processing/src/main/java/org/apache/druid/query/QueryMetrics.java:
```diff
@@ -236,6 +236,12 @@
   void identity(String identity);
 
+  /**
+   * Sets whether are not a segment scan has been vectorized. Generally expected to only be attached to segment-level
```
Whoa, how did I miss that? Rookie mistake.

gianm commented on processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java:
```diff
@@ -127,6 +123,59 @@ private GroupByQueryEngineV2()
                                     : DateTimes.utc(Long.parseLong(fudgeTimestampString));
 
+    final Filter filter = Filters.convertToCNFFromQueryContext(query, Filters.toFilter(query.getFilter()));
+    final Interval interval = Iterables.getOnlyElement(query.getIntervals());
```
Yeah, I figured that too.

gianm commented on processing/src/main/java/org/apache/druid/query/aggregation/DoubleSumVectorAggregator.java:
```diff
@@ -0,0 +1,83 @@
+
+    for (int i = 0; i < numRows; i++) {
+      final int position = positions[i] + positionOffset;
+      buf.putDouble(position, buf.getDouble(position) + vector[rows != null ? rows[i] : i]);
```
I'm not sure if it matters or not. It might be something interesting to check out after the dust settles on the patch. In the face of uncertainty about which path is best, I decided to err on the side of the more readable code. (Which I think the current patch is)

gianm commented on processing/src/main/java/org/apache/druid/segment/vector/VectorValueSelector.java:
```diff
@@ -0,0 +1,55 @@
+   * known that there are no nulls in the vector, possibly because the column is non-nullable.
+   */
+  @Nullable
+  boolean[] getNullVector();
```
Honestly I hadn't tried options other than the `boolean[]`. The amount of memory use here is minimal and I thought the array would be faster. But I didn't put a ton of thought into this decision. It could change later if we do some experiments to see what's actually best.

gianm commented on processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/ByteBufferHashTable.java:
```diff
@@ -254,18 +253,21 @@ protected void initializeNewBucketKey(
       final ByteBuffer keyBuffer,
-      final int keyHash
+      final int keyHash,
+      final Runnable preTableGrowthRunnable
```
Ah, thanks. Added it.

gianm commented on processing/src/main/java/org/apache/druid/query/aggregation/AggregatorAdapter.java:
```diff
@@ -0,0 +1,43 @@
+import java.io.Closeable;
+import java.nio.ByteBuffer;
+
+public interface AggregatorAdapter extends Closeable
```
I missed this comment. I just did it and added some more javadocs to AggregatorAdapters.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/QueryableIndexCursorSequenceBuilder.java:
```diff
@@ -0,0 +1,638 @@
+   * At this threshold, timestamp searches switch from binary to linear. See
+   * {@link #timeSearch(NumericColumn, long, int, int, int)} for more details.
+   */
+  private static final int TOO_CLOSE_FOR_MISSILES = 15000;
```
[![jester's dead](https://img.youtube.com/vi/JPgMyh7ei6w/0.jpg)](https://www.youtube.com/watch?v=JPgMyh7ei6w)

egor-ryashin replied:
Just wonder how that value was chosen?

gianm replied:
See this comment from the `timeSearch` method:

> The idea is to avoid too much decompression buffer thrashing. The default value `TOO_CLOSE_FOR_MISSILES` is chosen to be similar to the typical number of timestamps per block.

I moved the sentence about choice of default value to the javadoc for `TOO_CLOSE_FOR_MISSILES`, and kept the "idea" comment in the javadoc for `timeSearch`.

egor-ryashin replied:
It's not clear why `the typical number of timestamps per block` is 15000?

gianm replied:
> It's not clear why the typical number of timestamps per block is 15000?

Each block in a long typed column is 64KB, so if a timestamp can be stored in 3-4 bytes (typical with delta encoding) there's roughly that number per block. Or really a bit more.

But actually, you know what, I had never benchmarked it until today. I just benchmarked it and it looks like the pure binary search is fastest anyway. So, I removed the switching and made the search purely binary.

Benchmark results:

```
Benchmark                (query)  (rowsPerSegment)    (timestampString)  (tooCloseForMissiles)  (vectorize)  Mode  Cnt   Score    Error  Units
timeSearch                    15           5000000                 2000                      0        force  avgt   10   0.647 ±  0.001  ms/op
timeSearch                    15           5000000                 2000                   5000        force  avgt   10   0.643 ±  0.001  ms/op
timeSearch                    15           5000000                 2000                  15000        force  avgt   10   0.643 ±  0.001  ms/op
timeSearch                    15           5000000                 2000                  25000        force  avgt   10   0.579 ±  0.001  ms/op
timeSearch                    15           5000000                 2000                  50000        force  avgt   10   0.514 ±  0.001  ms/op
timeSearch                    15           5000000                 2000               99999999        force  avgt   10  ≈ 10⁻⁴           ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00                      0        force  avgt   10   0.578 ±  0.001  ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00                   5000        force  avgt   10   0.589 ±  0.001  ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00                  15000        force  avgt   10   0.588 ±  0.001  ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00                  25000        force  avgt   10   0.751 ±  0.008  ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00                  50000        force  avgt   10   0.607 ±  0.004  ms/op
timeSearch                    15           5000000  2000-01-01T01:00:00               99999999        force  avgt   10   2.106 ±  0.078  ms/op
timeSearch                    15           5000000     2000-01-01T10:30                      0        force  avgt   10   0.708 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T10:30                   5000        force  avgt   10   0.722 ±  0.002  ms/op
timeSearch                    15           5000000     2000-01-01T10:30                  15000        force  avgt   10   0.728 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T10:30                  25000        force  avgt   10   0.773 ±  0.016  ms/op
timeSearch                    15           5000000     2000-01-01T10:30                  50000        force  avgt   10   0.854 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T10:30               99999999        force  avgt   10  21.863 ±  0.146  ms/op
timeSearch                    15           5000000     2000-01-01T16:30                      0        force  avgt   10   0.643 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T16:30                   5000        force  avgt   10   0.653 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T16:30                  15000        force  avgt   10   0.668 ±  0.001  ms/op
timeSearch                    15           5000000     2000-01-01T16:30                  25000        force  avgt   10   0.685 ±  0.004  ms/op
timeSearch                    15           5000000     2000-01-01T16:30                  50000        force  avgt   10   0.935 ±  0.013  ms/op
timeSearch                    15           5000000     2000-01-01T16:30               99999999        force  avgt   10  39.498 ±  0.022  ms/op
timeSearch                    15           5000000           2000-01-02                      0        force  avgt   10   0.602 ±  0.001  ms/op
timeSearch                    15           5000000           2000-01-02                   5000        force  avgt   10   0.616 ±  0.003  ms/op
timeSearch                    15           5000000           2000-01-02                  15000        force  avgt   10   0.645 ±  0.001  ms/op
timeSearch                    15           5000000           2000-01-02                  25000        force  avgt   10   0.707 ±  0.001  ms/op
timeSearch                    15           5000000           2000-01-02                  50000        force  avgt   10   0.811 ±  0.001  ms/op
timeSearch                    15           5000000           2000-01-02               99999999        force  avgt   10  61.194 ±  0.036  ms/op
```

gianm replied:
@egor-ryashin What do you think of the latest changes?

egor-ryashin replied:
Removing **tooCloseForMissiles** check, isn't it an equivalent of setting `tooCloseForMissiles = 99999999 ` 
 which shows slower operations?

gianm replied:
@egor-ryashin It would be the equivalent of tooCloseForMissiles = 0, since I removed a check that did a “break” out of the binary search when maxIndex and minIndex were less than tooCloseForMissiles apart.

 (the threshold was: when search is this close, switch to linear. So 99999999 means fully linear and 0 means fully binary)

gianm commented:
Going to merge in #8049 to this branch when it's available, since that should clean up Travis results again. Thanks to everyone that reviewed. @egor-ryashin I think I have addressed all your comments, let me know if you agree. I'm looking to merge this to master soon so we can start building on top of it.

gianm commented:
Just pushed up another commit with this change: https://github.com/apache/incubator-druid/pull/6794#discussion_r302651747

gianm commented:
Thanks everyone!!
